plugins/logger/filesystem_logger.cpp

```diff
@@ -32,35 +34,179 @@ FLAG(string,
 /// Legacy, backward compatible "osquery_log_dir" CLI option.
 FLAG_ALIAS(std::string, osquery_log_dir, logger_path);
 
+FLAG(bool, logger_rotate, false, "Use filesystem log rotation");
+FLAG(uint64,
+     logger_rotate_max_files,
+     25,
+     "Max number of files to keep in rotation");
+FLAG(uint64,
+     logger_rotate_size,
+     25 * 1024 * 1024,
+     "Size for each filesystem log in bytes");
+
 FLAG(int32, logger_mode, 0640, "Decimal mode for log files (default '0640')");
 
 const std::string kFilesystemLoggerFilename = "osqueryd.results.log";
 const std::string kFilesystemLoggerSnapshots = "osqueryd.snapshots.log";
 
+bool LogRotate::shouldRotate() {
+  if (this->fileSize(path_) >= this->getRotateSize()) {
+    return true;
+  }
+  return false;
+}
+
+Status LogRotate::rotate(size_t max_files) {
+  if (!pathExists(path_)) {
+    return Status::failure("Path under rotation does not exist");
+  }
+  size_t offset = 1;
+  std::vector<std::string> moves{getRotateFile(offset)};
+  while (this->pathExists(moves.back())) {
+    moves.push_back(getRotateFile(++offset));
+  }
+
+  auto s = this->moveFiles(moves, max_files);
+  if (!s.ok()) {
+    return s;
+  }
+  return this->moveFile(path_, *moves.begin());
+}
+
+size_t LogRotate::getRotateSize() {
+  return FLAGS_logger_rotate_size;
+}
+
+size_t LogRotate::fileSize(const std::string& filepath) {
+  boost::system::error_code ec;
+  auto size = fs::file_size(filepath, ec);
+  return (ec) ? 0 : size;
+}
+
+bool LogRotate::pathExists(const std::string& path) {
+  return osquery::pathExists(path).ok();
+}
+
+Status LogRotate::removeFile(const std::string& path) {
+  return osquery::removePath(path);
+}
+
+Status LogRotate::moveFile(const std::string& source, const std::string& dest) {
+  return osquery::movePath(source, dest);
+}
+
+Status LogRotate::compressFile(const std::string& source,
+                               const std::string& dest) {
+  return osquery::compress(source, dest);
+}
+
+std::string LogRotate::getRotateFile(size_t offset) {
+  auto filename = path_ + '.' + std::to_string(offset);
+  if (offset > kUncompressedCount) {
+    filename += ".zst";
```

Comment:
This seems like a fairly uncommon compression format. Though it has the advantage of already being in our codebase

Reply:
Yeap, this is again me reaching for what exists and not over-engineering or setting new prescient. My understanding is one wants to keep these archives around for local debugging but your aggregation solution will only look at the `.1` non-compressed version. So I think it's acceptable to have something slightly non-standard, through z _standard_ is gaining popularity.

Reply:
Reasonable enough. We can get something out, and if there's pushback, we can change it